# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [1]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/45/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/45/"
    },
    "https://swapi.dev/api/people/60/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/60/"
    },
    "https://swapi.dev/api/planets/1/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/1/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1:
- Integrante 2:
- Integrante 3 (se houver):
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 45

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [3]:
url = "https://swapi.dev/api/people/45/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Bib Fortuna",
    "height": "180",
    "mass": "unknown",
    "hair_color": "none",
    "skin_color": "pale",
    "eye_color": "pink",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/37/",
    "films": [
        "https://swapi.dev/api/films/3/"
    ],
    "species": [
        "https://swapi.dev/api/species/15/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T09:47:02.512000Z",
    "edited": "2014-12-20T21:17:50.407000Z",
    "url": "https://swapi.dev/api/people/45/"
}


**Respostas do aluno — Personagem 45**

- Status code:200
- Nome do personagem: "Bib Fortuna"
- Altura: "180"
- Cor dos olhos:"pink",
- O JSON parece completo? sim

### 1.2 Personagem 77

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [5]:
url = "https://swapi.dev/api/people/77/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "San Hill",
    "height": "191",
    "mass": "unknown",
    "hair_color": "none",
    "skin_color": "grey",
    "eye_color": "gold",
    "birth_year": "unknown",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/57/",
    "films": [
        "https://swapi.dev/api/films/5/"
    ],
    "species": [
        "https://swapi.dev/api/species/34/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-20T17:58:17.049000Z",
    "edited": "2014-12-20T21:17:50.484000Z",
    "url": "https://swapi.dev/api/people/77/"
}


**Respostas do aluno — Personagem 77**

- Status code: 200
- Nome do personagem:"San Hill"
- Altura: "191"
- Cor dos olhos: "gold"
- O JSON parece completo? sim

### 1.3 Planeta 17

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [6]:
url = "https://swapi.dev/api/planets/17/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Felucia",
    "rotation_period": "34",
    "orbital_period": "231",
    "diameter": "9100",
    "climate": "hot, humid",
    "gravity": "0.75 standard",
    "terrain": "fungus forests",
    "surface_water": "unknown",
    "population": "8500000",
    "residents": [],
    "films": [
        "https://swapi.dev/api/films/6/"
    ],
    "created": "2014-12-10T13:44:50.397000Z",
    "edited": "2014-12-20T20:58:18.447000Z",
    "url": "https://swapi.dev/api/planets/17/"
}


**Respostas do aluno — Planeta 17**

- Status code: 200.
- Nome do planeta: "Felucia".
- Clima: "hot, humid", ou seja quente e húmido.
- Terreno:"fungus forests", ou seja Floresta de Fungos.
- População: "8500000"
- O JSON parece completo? sim

### 1.4 Nave 29

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [9]:
url = "https://swapi.dev/api/starships/29/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "B-wing",
    "model": "A/SF-01 B-wing starfighter",
    "manufacturer": "Slayn & Korpil",
    "cost_in_credits": "220000",
    "length": "16.9",
    "max_atmosphering_speed": "950",
    "crew": "1",
    "passengers": "0",
    "cargo_capacity": "45",
    "consumables": "1 week",
    "hyperdrive_rating": "2.0",
    "MGLT": "91",
    "starship_class": "Assault Starfighter",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-18T11:18:04.763000Z",
    "edited": "2014-12-20T21:23:49.909000Z",
    "url": "https://swapi.dev/api/starships/29/"
}


**Respostas do aluno — Nave 29**

- Status code: 200
- Nome da nave: "B-wing"
- Modelo: "A/SF-01 B-wing starfighter"
- Fabricante: "Slayn & Korpil"
- Velocidade máxima em atmosfera: "950"
- O JSON parece completo? sim

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [10]:
url = "https://swapi.dev/api/people/19/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Jek Tono Porkins
Altura: 180
Cor dos olhos: blue
Gênero: male


**Perguntas**
1. Qual campo foi usado para mostrar o nome? print("Nome:", dados["name"])
2. Qual campo foi usado para mostrar a altura? print("Altura:", dados["height"])
3. Se você fosse QA, quais desses campos validaria primeiro? Eu validaria o do Nome primeiro.

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [17]:
url = "https://swapi.dev/api/planets/19/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Saleucami
Clima: hot
Terreno: caves, desert, mountains, volcanoes
População: 1400000000


**Perguntas**
1. O campo população parece numérico? Sim.
2. O clima foi retornado corretamente? Sim,retornou que o clima predominante e muito quente, ou seja de muito calor na atmosfera.
3. Se você fosse QA, o que validaria nesse planeta? Eu validaria o clima, pois nome tem varios, agora o clima é muito específico.

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [18]:
url = "https://swapi.dev/api/people/31/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? Verifica o Status Code, ou seja se há uma conexão com o banco de dados.
2. O que a validação 2 verifica? Verifica a existencia de um nome em letras.
3. O que a validação 3 verifica? Verifica se o campo nome não está null, ou seja se não está vazio.
4. O que a validação 4 verifica? Verifica se possui dados complementares.

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [19]:
url = "https://swapi.dev/api/starships/22/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos. //// já foi alterado e continua ok.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [21]:
url = "https://swapi.dev/api/people/1/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


Respostas

1º Linha: Confirma que a requisição HTTP funcionou.

2ª Linha: Confirma que o JSON retornado possui o campo name.

3ª Linha: Confirma que o campo name não está vazio.

4ª Linha: Essa mensagem só aparece se todos os asserts forem verdadeiros.

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [23]:
url = "https://swapi.dev/api/starships/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


Respostas.

1º Linha: Confirma que a requisição HTTP funcionou.

2ª Linha: Confirma que o JSON retornado possui o campo fabricante da nave.

3ª Linha: Confirma que o campo modelo da nave não está vazio.

4ª Linha: Essa mensagem só aparece se todos os asserts forem verdadeiros.


*************************SUGESTÃO **************


Uma validação adicional poderia verificar se o campo name da nave existe e não está vazio.

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [ ]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? 404
2. Esse comportamento está correto? Sim, por que o valor informado foi inexistente.
3. Por que um QA deve testar cenários negativos? Para a fluides e a confiança 100% no aplicativo ou software desenvolvido.

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [35]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/8/")

Origem dos dados: online
Status Code: 200
name: Naboo
rotation_period: 26
orbital_period: 312
diameter: 12120
climate: temperate
gravity: 1 standard
terrain: grassy hills, swamps, forests, mountains
surface_water: 12
population: 4500000000
residents: ['https://swapi.dev/api/people/3/', 'https://swapi.dev/api/people/21/', 'https://swapi.dev/api/people/35/', 'https://swapi.dev/api/people/36/', 'https://swapi.dev/api/people/37/', 'https://swapi.dev/api/people/38/', 'https://swapi.dev/api/people/39/', 'https://swapi.dev/api/people/42/', 'https://swapi.dev/api/people/60/', 'https://swapi.dev/api/people/61/', 'https://swapi.dev/api/people/66/']
films: ['https://swapi.dev/api/films/3/', 'https://swapi.dev/api/films/4/', 'https://swapi.dev/api/films/5/', 'https://swapi.dev/api/films/6/']
created: 2014-12-10T11:52:31.066000Z
edited: 2014-12-20T20:58:18.430000Z
url: https://swapi.dev/api/planets/8/


**Respostas finais**
1. Qual endpoint você testou no desafio? resumo_recurso("https://swapi.dev/api/planets/8/") ////no caso alterei para o numero 8///.

2. Qual foi o status code? 200.

3. Qual campo chamou mais a sua atenção? population: 4500000000.

4. Se você fosse QA, quais 3 validações faria nesse recurso?

(1) Status Code: 200

(2) name: Naboo

(3) url: https://swapi.dev/api/planets/8/.

# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
